src/meta/processors/admin/Balancer.cpp

```diff
@@ -78,14 +89,18 @@ bool Balancer::recovery() {
         CHECK_EQ(1, corruptedPlans.size());
         plan_ = std::make_unique<BalancePlan>(corruptedPlans[0], kv_, client_.get());
         plan_->onFinished_ = [this] () {
-            bool expected = true;
-            auto &running = this->running_;   // Get the reference before the captured `this' lost
-            plan_.reset();
-            CHECK(running.compare_exchange_strong(expected, false));
+            // Get the reference before the captured `this' lost
+            auto &lock = this->lock_;
+            {
+                std::lock_guard<std::mutex> lg(lock);
+                running_ = false;
+                plan_.reset();
+            }
         };
         if (!plan_->recovery()) {
             LOG(ERROR) << "Can't recovery plan " << corruptedPlans[0];
-            plan_->onFinished_();
+            running_ = false;
```

Comment:
why not call onFinished_ directly?

Reply:
I has a lock in onFinished_, and recovery has been locked outside.